nx_parallel/algorithms/shortest_paths/dense.py

```diff
@@ -0,0 +1,251 @@
+                            i_range = (block_i * blocking_factor, n - 1)
+                        if block_j == no_of_primary - 1:
+                            j_range = (block_j * blocking_factor, n - 1)
+                    params.append((i_range, j_range))
+        Parallel(n_jobs=(no_of_primary - 1) ** 2, require="sharedmem")(
+            delayed(_partial_floyd_warshall_numpy)(A, k, i, j) for (i, j) in params
+        )
+    dist = _matrix_to_dict(A, nodelist)
+
+    return dist
+
+
+def _partial_floyd_warshall_numpy(A, k_iteration, i_iteration, j_iteration):
+    """
+    Compute partial FW in the determined sub-block for the execution of
+    parallel tiled FW.
+
+    Parameters
+    ----------
+    A : 2D numpy.ndarray
+        matrix that reppresent the adjacency matrix of the graph
+
+    k_iteration : tuple
+        range (start-end) of the primary block in the current iteration
+
+    i_iteration : tuple
+        range (start-end) of the rows in the current block computed
+
+    j_iteration : tuple
+        range (start-end) of the columns in the current block computed
+
+    Returns
+    -------
+    A : 2D numpy.ndarray
+        adjacency matrix updated after floyd warshall
+    """
+    for k in range(k_iteration[0], k_iteration[1] + 1):
+        for i in range(i_iteration[0], i_iteration[1] + 1):
+            for j in range(j_iteration[0], j_iteration[1] + 1):
+                A[i][j] = min(A[i][j], (A[i][k] + A[k][j]))
+
+
+def _block_range(blocking_factor, block):
+    return (block * blocking_factor, (block + 1) * blocking_factor - 1)
+
+
+def _calculate_divisor(i, x, y):
+    if x % i == 0:
+        divisor1 = result2 = i
+        result1 = divisor2 = x // i
+        # difference1 = abs((result1 - 1) ** 2 - y)
+        difference1 = abs(result1 - y)
+        # difference2 = abs((result2 - 1) ** 2 - y)
+        difference2 = abs(result2 - y)
+
+        if difference1 < difference2:
+            return divisor1, result1, difference1
+        else:
+            return divisor2, result2, difference2
+    return None
+
+
+def _find_nearest_divisor(x, y):
+    """
+    find the optimal value for the blocking factor parameter
+
+    Parameters
+    ----------
+    x : node number
+
+    y : cpu core available
+    """
+    if x < y:
+        return 1, False
+    # Find the square root of x
+    sqrt_x = int(math.sqrt(x)) + 1
+
+    # Execute the calculation in parallel
+    results = Parallel(n_jobs=-1)(
+        delayed(_calculate_divisor)(i, x, y) for i in range(2, sqrt_x)
+    )
+
+    # Filter out None results
+    results = [r for r in results if r is not None]
+
+    if len(results) <= 0:
+        # This check if a number is prime, although repeat process with a non prime number
+        best_divisor, _ = _find_nearest_divisor(x - 1, y)
+        return best_divisor, True
+    # Find the best divisor
+    best_divisor, _, _ = min(results, key=lambda x: x[2])
+
+    return best_divisor, False
```

Comment:
Curious-- for large graphs with prime numbers of nodes, it seems like the final block could be disproportionately larger, leading to load imbalance across cores?

Reply:
Hm, what about just swapping:
```
best_divisor, _ = _find_nearest_divisor(x - 1, y)
```
with something as simple as:
```
best_divisor = max(1, x // y)
```
to distribute load approximately evenly if no divisor is found?